### PR TITLE
Add timeouts to JS SDK

### PR DESCRIPTION
**Pull request by jakubno**

### CODE REVIEW

File: packages/js-sdk/src/session/filesystem.ts [hunk 1]
```diff
@@ -1,4 +1,5 @@
 import FilesystemWatcher from './filesystemWatcher'
+import { CallOpts } from './sessionConnection.ts'
```
Comment (Strajk) [inline packages/js-sdk/src/session/filesystem.ts]: TS is (correctly) complaining about the `.ts` suffix

```suggestion
import { CallOpts } from './sessionConnection'
```

image

File: packages/js-sdk/src/session/index.ts [hunk 2]
```diff
@@ -40,34 +52,42 @@ export class Session extends SessionConnection {
     this.filesystem = {
       /**
        * List files in a directory.
-       * @param path path to a directory
+       * @param path P}ath to a directory
```
Comment (Strajk) [inline packages/js-sdk/src/session/index.ts]:
```suggestion
       * @param path Path to a directory
```

typo 🙈

File: packages/js-sdk/src/session/filesystemWatcher.ts [hunk 3]
```diff
@@ -32,18 +33,21 @@ class FilesystemWatcher {
   }
 
   // Starts watching the path that was passed to the contructor
-  async start() {
-    // Already started.
-    if (this.rpcSubscriptionID) return
+  async start(opts: CallOpts) {
+    const _start = async () => {
```
Comment (Strajk) [inline packages/js-sdk/src/session/filesystemWatcher.ts]: Nitpick: Noticed inconsistent naming of "original" functions 
`start > _start` vs `open > open`
Screen 2023-09-04 at 15 47 07

File: packages/js-sdk/src/session/process.ts [hunk 4]
```diff
@@ -100,20 +100,24 @@ export class Process {
...
+   * @param opts Call options
+   * @param {timeout} [opts.timeout] Timeout in milliseconds
    */
-  async sendStdin(data: string): Promise<void> {
-    await this.session.call(processService, 'stdin', [this.processID, data])
+  async sendStdin(data: string, opts?: CallOpts): Promise<void> {
+    await this.session.call(processService, 'stdin', [this.processID, data], opts)
   }
 }
+export interface ProcessProps {
```
Comment (Strajk) [inline packages/js-sdk/src/session/process.ts]: maybe `ProcessOpts` would be more consistent within the codebase, `Props` sounds too "React"y

Screen 2023-09-04 at 15 52 15

File: packages/js-sdk/src/utils/promise.ts [hunk 5]
```diff
@@ -37,3 +40,21 @@ export function createDeferredPromise<T = void>() {
...
+  fn: T,
+  timeout: number = TIMEOUT,
+): T {
+  if (timeout === undefined || timeout <= 0 || timeout === Number.POSITIVE_INFINITY) {
+    return fn
+  }
+
+  // Throw an error if it takes too long
+  const timer = new Promise((resolve, reject) => {
+    setTimeout(() => reject(new TimeoutError('The process took too long.')), timeout)
```
Comment (Strajk) [inline packages/js-sdk/src/utils/promise.ts]: What about sth like: `Calling "${fn.name}" timeouted after ${timeout}ms`

File: packages/js-sdk/src/session/terminal.ts [hunk 6]
```diff
@@ -64,24 +64,30 @@ export class Terminal {
...
-    /**
-     * Working directory where will the terminal start.
-     */
-    rootdir?: string
-    /**
-     * Environment variables that will be accessible inside of the terminal.
-     */
-    envVars?: EnvVars
-  }) => Promise<Terminal>
+  readonly start: (opts: TerminalProps) => Promise<Terminal>
```
Comment (Strajk) [inline packages/js-sdk/src/session/terminal.ts]: hmmm, dejavu... what about calling these TerminalOpts?

Comment (jakubno) [reply packages/js-sdk/src/session/terminal.ts]: I have a feeling I was searching for props in the project and was renaming it. Now I am confused

**Review by Strajk — APPROVED** [review]
Haven't thoroughly checked but looks good 👍

**Pull request merged**